### DISCUSSION TITLE
Difference between obsonly = T and obsonly = F - can someone explain to a newbie

### DISCUSSION THREAD

**Discussion by algaloo** [metrumresearchgroup/mrgsolve#995] [General]
Hi.  I know what the stated difference is between obsonly = T and obsonly = F in a simulation, but I don't understand it.

code <- '
$PARAM TVCL = 1.23, TVV = 35.7, TVKA = 1.3
F1 = 0.82, ALAG = 1.21
WT = 70, SEX = 0

$MAIN
double CL = TVCL*pow(WT/70,0.75)*exp(ECL);
double V  = TVV*(WT/70)*exp(EV);
double KA = TVKA*exp(EKA);

if(SEX==1) V = V*0.8;

F_GUT = F1;
ALAG_GUT = ALAG;

$PKMODEL cmt="GUT CENT", depot=TRUE

$OMEGA @labels ECL EV EKA
0.015 0.2 0.5

$SIGMA @labels PROP ADD
0.03 230

$TABLE
capture IPRED = CENT/(V/1000);
capture DV = IPRED*(1+PROP)+ADD;

while(DV < 0) {
  simeps();
  DV = IPRED*(1+PROP)+ADD;
}

$CAPTURE WT CL DV IPRED
'

mod <- mcode_cache("demo", code)

set.seed(33020)
idata <- 
  data_frame(ID=1:2000) %>% 
  mutate_random(WT[50,110] ~ rnorm(80,30)) %>% 
  mutate_random(SEX ~ rbinomial(0.7)) %>%
  mutate(dosegr = as.integer(WT > 90))

ev1 <- ev(amt=100, ii=24, addl=9)
ev2 <- ev(amt=150, ii=24, addl=9)

data <- assign_ev(list(ev1,ev2),idata,"dosegr")

set.seed(11009)

out <- 
  mod %>% 
  data_set(data) %>%
  idata_set(idata) %>%
  carry_out(dosegr) %>%
  mrgsim(delta=1, end=360, **obsonly=TRUE**)

If I change the boldface to **obsonly=FALSE** I don't quite see what the difference is.

**Comment by kylebaron** [comment]
Hi @algaloo  - 

Here's the difference (slightly modified so we can show it):

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(mrgsolve)
#> 
#> Attaching package: 'mrgsolve'
#> The following object is masked from 'package:stats':
#> 
#>     filter
library(dmutate)

code <- '
$PARAM TVCL = 1.23, TVV = 35.7, TVKA = 1.3
F1 = 0.82, ALAG = 1.21
WT = 70, SEX = 0

$MAIN
double CL = TVCL*pow(WT/70,0.75)*exp(ECL);
double V = TVV*(WT/70)*exp(EV);
double KA = TVKA*exp(EKA);

if(SEX==1) V = V*0.8;

F_GUT = F1;
ALAG_GUT = ALAG;

$PKMODEL cmt="GUT CENT", depot=TRUE

$OMEGA @labels ECL EV EKA
0.015 0.2 0.5

$SIGMA @labels PROP ADD
0.03 230

$TABLE
capture IPRED = CENT/(V/1000);
capture DV = IPRED*(1+PROP)+ADD;

while(DV < 0) {
simeps();
DV = IPRED*(1+PROP)+ADD;
}

$CAPTURE WT CL DV IPRED
'

mod <- mcode_cache("demo", code)
#> Building demo ...
#> done.

set.seed(33020)
idata <-
  data.frame(ID=1:2000) %>%
  mutate_random(WT[50,110] ~ rnorm(80,30)) %>%
  mutate_random(SEX ~ rbinomial(0.7)) %>%
  mutate(dosegr = as.integer(WT > 90))
#> Warning: `data_frame()` was deprecated in tibble 1.1.0.
#> Please use `tibble()` instead.
#> This warning is displayed once every 8 hours.
#> Call `lifecycle::last_lifecycle_warnings()` to see where this warning was generated.

ev1 <- ev(amt=100, ii=24, addl=9)
ev2 <- ev(amt=150, ii=24, addl=9)

data <- assign_ev(list(ev1,ev2),idata,"dosegr")

set.seed(11009)

out_oo_true <-
  mod %>%
  data_set(data) %>%
  idata_set(idata) %>%
  carry_out(dosegr, evid) %>%
  mrgsim_df(delta=1, end=360, obsonly=TRUE)

out_oo_false <-
  mod %>%
  data_set(data) %>%
  idata_set(idata) %>%
  carry_out(dosegr, evid) %>%
  mrgsim_df(delta=1, end=360, obsonly=FALSE)


dim(out_oo_true)
#> [1] 722000     10
dim(out_oo_false)
#> [1] 724000     10

count(out_oo_true, evid)
#>   evid      n
#> 1    0 722000
count(out_oo_false, evid)
#>   evid      n
#> 1    0 722000
#> 2    1   2000
```

<sup>Created on 2022-06-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>
